### ISSUE TITLE
[APPENG-6062] ML-based Mantis RCA classification upstream of ran-rca-service

### ISSUE THREAD

**Issue by TomerG2** [rh-ai-quickstart/ai-driven-network-remediation#146]
## Problem Statement

The Telco/O-RAN workflow today enriches rule-detected RAN anomalies using **LLM-only** root cause analysis in `ran-rca-service` (`RAG Retrieval` → `Analyze`). Operators and demo audiences see Granite explain failures and cite vendor documentation, but there is no **ML-based root-cause classification** stage — even though fine-tuned Mantis-8M work (TelecomTS, 10 RCA classes) exists and OpenShift AI model serving is a core platform story for this quickstart.

Without an ML upstream stage, the agent cannot demonstrate the hybrid **train → serve on OpenShift AI → consume in LangGraph** loop, and the LLM must infer root-cause category from symptom text alone (`LowRsrp`, etc.) rather than from KPI time-series patterns the model was trained on.

## Solution

Add **ML-based anomaly classification** (fine-tuned Mantis-8M) as an upstream graph stage in `ran-rca-service`, served via a **hybrid predictor service** (`ran-ml-service`) that runs locally for dev/CI and on **OpenShift AI model serving (KServe InferenceService)** in the cluster.

The detector publishes a **KPI window snapshot** on `ran-anomalies` when an anomaly fires. The new **Classify** node calls the predictor; when ML is available and confident enough, the predicted RCA class steers **RAG Retrieval** and **Analyze**; the LLM still produces explanation and vendor-doc-grounded **recommended fix**. Enriched Kafka records and the RAN chatbot surface both ML prediction and LLM conclusions, including when ML steering was skipped.

## User Stories

1. As a **network operator**, I want ML to classify a detected anomaly into one of 10 known root-cause categories, so that RCA is grounded in KPI patterns rather than symptom text alone.
2. As a **network operator**, I want the LLM to still explain the root cause and recommend a vendor-documented fix, so that I get actionable remediation guidance.
3. As a **network operator**, I want low-confidence ML predictions to not steer retrieval or analysis, so that uncertain classifications do not mislead the LLM.
4. As a **network operator**, I want to see ML prediction and confidence in chat context even when steering was skipped, so that I can compare what ML suggested versus what the LLM concluded.
5. As a **platform engineer**, I want the Mantis predictor deployed as an InferenceService on OpenShift AI, so that the demo showcases RH model-serving capabilities.
6. As a **platform engineer**, I want the same predictor image to run via `uv run` locally, so that dev and CI do not require a full cluster.
7. As a **platform engineer**, I want the predictor to load weights from `MLFLOW_MODEL_URI` or a local path, so that MLflow registry integration can be demonstrated without blocking initial wiring.
8. As a **data scientist**, I want a single shared `RCA_CLASSES` vocabulary in the repo, so that training labels, predictor output, enriched records, and tests do not drift.
9. As a **data scientist**, I want feature normalization and tensor shaping confined to the predictor, so that the detector remains telemetry-focused.
10. As a **SRE**, I want anomalies without a sufficient KPI window to fall back to LLM-only RCA, so that cold-start and partial history do not block the pipeline.
11. As a **SRE**, I want enriched records to always include ML fields (empty/zero when skipped), so that downstream consumers have a stable schema.
12. As a **integration test author**, I want JSON Schema validation on enriched output, so that contract breaks are caught in CI.
13. As a **demo presenter**, I want a clear narrative: detector → ML classify on OAI → LangGraph RAG + LLM → enriched topic → chatbot, so that the end-to-end AI stack is visible.
14. As a **consumer of `ran-anomalies`**, I want `kpi_window` to be optional during rollout, so that detector and RCA service can deploy independently.
15. As a **chatbot user**, I want the assistant to note when ML steering was not used, so that low-confidence predictions are transparent.
16. As a **compliance reviewer**, I want `ml_steer_used` on enriched records, so that audit trails distinguish ML-guided versus LLM-only analysis.
17. As a **Helm operator**, I want InferenceService and predictor Deployment wired in the hub chart, so that cluster deploy is one-step with other services.
18. As a **developer**, I want `MANTIS_CONFIDENCE_THRESHOLD` configurable via env, so that demo tuning does not require code changes.
19. As a **developer**, I want `MANTIS_MIN_WINDOW` aligned to model `seq_len`, so that window length matches training expectations.
20. As a **developer**, I want detector `history_size` aligned to `MANTIS_MIN_WINDOW`, so that published windows are usually sufficient after warm-up.
21. As a **downstream analytics consumer**, I want `ml_confidence` on every enriched message, so that model quality can be monitored over time.
22. As a **maintainer**, I want classify node failures to degrade to LLM-only RCA, so that predictor outages do not stall the Kafka pipeline.

### Pipeline and graph

- Extend `ran-rca-service` LangGraph pipeline: **`classify` → `rag_retrieval` → `analyze`** (ADR-0001: state remains Pydantic `BaseModel`).
- **Classify** calls `MANTIS_INFERENCE_URL` (`POST /v1/classify` with `kpi_window`). On skip/failure, state proceeds without ML steering fields populated beyond empty defaults.
- **RAG Retrieval** query when `ml_steer_used` is true: `{ml_root_cause_class} {anomaly_type} {anomaly} {band}`. Otherwise unchanged: `{anomaly_type} {anomaly} {band}`.
- **Analyze** prompt includes ML class hint when `ml_steer_used` is true; otherwise unchanged.

### KPI window on `ran-anomalies` (detector change)

- `ran-anomaly-detector` publishes optional `kpi_window` on each anomaly output, snapshot from `MetricsStore` at detection time for the firing `(cell_id, band)`.
- **`kpi_window` shape** — list of flat per-timestep dicts (no nested cell, no tensors):

```json
{
  "timestamp": "2026-08-19T10:08:10Z",
  "rsrp": -125.0,
  "rsrq": -19.0,
  "sinr": -4.1,
  "throughput_mbps": 6.2,
  "ues_usage": 22,
  "latency_ms": 45.7
}
```

- Parent message retains `cell_id`, `band`, `anomaly_type`, `anomaly`.
- Align detector `history_size` with `MANTIS_MIN_WINDOW` (from model artifact `seq_len`, default 10).

### ML skip and confidence gating

- **Skip ML** (LLM-only path for RAG/analyze steering) when:
  - `kpi_window` is missing or empty
  - `len(kpi_window) < MANTIS_MIN_WINDOW`
  - Predictor call fails
- **Confidence gate** (`MANTIS_CONFIDENCE_THRESHOLD`, default **0.6**):
  - Below threshold: do not steer RAG or analyze; still record ML prediction for audit
  - Above threshold: set `ml_steer_used: true` and steer RAG + analyze

### Enriched Kafka output (`ran-anomalies-enriched`)

Always present fields (empty/zero/false when ML did not run or did not steer):

```json
{
  "ml_root_cause_class": "",
  "ml_confidence": 0.0,
  "ml_steer_used": false
}
```

When ML ran but was gated: `ml_root_cause_class` and `ml_confidence` populated, `ml_steer_used: false`.

### Shared RCA class vocabulary

- Define **`RCA_CLASSES`** (10 TelecomTS labels) in **`telco-oran`** as single source of truth.
- Predictor maps class index → `RCA_CLASSES[index]`; invalid index → ML failure → LLM-only fallback.
- Labels (order matters for index mapping):

  1. Antenna Failure
  2. Co-Channel Interference (Mild)
  3. Co-Channel Interference (Severe)
  4. Faulty RF Filters (Temporal)
  5. Doppler Shift (Severe)
  6. Faulty Handover Algorithm (Too Frequent)
  7. Buffer Overflow (Gradual Buildup)
  8. Resource Allocation Bugs
  9. High Network Congestion (Gradual Buildup)
  10. High Network Congestion (Sudden Spike)

### Hybrid predictor service (`ran-ml-service`)

- New hub service: FastAPI + PyTorch + Mantis.
- Endpoints: `POST /v1/classify`, `GET /health`, `GET /ready` (model loaded).
- Request: `{ "kpi_window": [...] }`. Response: `{ "class": "<RCA_CLASSES label>", "confidence": <float>, "class_index": <int> }` (class_index optional but useful for debug).
- **Model load**: support `MLFLOW_MODEL_URI` (demo/default path) and `MANTIS_MODEL_PATH` (CI/local fixture). Gate with `MANTIS_ENABLED`.
- Feature order, normalization, pad/truncate → tensor **only inside predictor** (align with Alan’s TelecomTS export).
- **Same OCI image** for local `uv run` and cluster InferenceService.

### OpenShift AI deployment (hybrid cluster path)

- Helm: InferenceService deploying the predictor image (custom container, port 8080, `/v1/classify`).
- `ran-rca-service` env: `MANTIS_INFERENCE_URL` → in-cluster InferenceService URL.
- Minimal KServe CRs in hub chart — enough to demo OAI model serving without requiring stock TGI/vLLM runtimes.

### RAN chatbot

- Extend `EnrichedAnomaly` model to parse new ML fields (defaults for backward compatibility during development).
- Update `_format_anomalies` to show ML prediction + confidence; when `ml_steer_used` is false and `ml_root_cause_class` is non-empty, note that steering was skipped (e.g. low confidence).

### Contracts

- Update `ran-anomaly-enriched.schema.json` with three ML fields.
- Add `ran-anomalies.schema.json` documenting optional `kpi_window` (recommended for implementers).

### Modules touched (summary)

| Module | Change |
|--------|--------|
| `telco-oran` | `RCA_CLASSES` constant |
| `ran-anomaly-detector` | Publish `kpi_window`; align `history_size` |
| `ran-ml-service` | **New** predictor service |
| `ran-rca-service` | Classify node; graph wiring; enriched projection; config |
| `ran-chatbot-service` | Parse + format ML fields in chat context |
| `hub/helm` | Predictor Deployment/InferenceService; env for RCA + ML URLs |
| `contracts` | Enriched + anomalies schemas |
| `Makefile` | Build/test targets for `ran-ml-service` |

### What makes a good test

- Test **observable behavior** at service boundaries: HTTP responses, graph invoke output, Kafka-shaped dicts, JSON Schema conformance.
- Do **not** test internal tensor math beyond smoke-level predictor tests with fixture weights.
- Prefer **existing seams** over new ones; minimize cross-service integration in unit tests.

### Primary test seam (highest — target one main seam)

**`ran-rca-service` graph end-to-end** via `build_graph().ainvoke(anomaly_dict)` — extend existing `test_graph.py` pattern:

- Mock predictor HTTP (or patch classify client) and existing RAG/LLM mocks.
- Assert enriched projection fields including ML columns for scenarios: ML steered, ML gated (low confidence), ML skipped (no window), predictor failure.
- Validate output against updated `ran-anomaly-enriched.schema.json` via `jsonschema` (prior art in `test_graph.py`).

### Secondary seams (narrower, only where graph tests cannot cover)

| Seam | What to assert |
|------|----------------|
| **`ran-ml-service` HTTP API** | `/v1/classify` with fixture `kpi_window` returns valid class label + confidence; `/ready` after load |
| **`ran-anomaly-detector` output** | Anomaly dict includes `kpi_window` with correct length and KPI keys after sufficient history (extend `test_detection.py` shape assertions) |
| **`ran-chatbot-service` formatting** | `_format_anomalies` includes ML line and steering note (unit test on formatter, no live Kafka) |

### Prior art

- `hub/ran-rca-service/tests/test_graph.py` — full graph invoke + schema validation
- `hub/ran-anomaly-detector/tests/test_detection.py` — output shape assertions
- `hub/ran-rca-service/tests/test_server.py` — Kafka handler enriched projection
- `hub/ran-chatbot-service/tests/` — chat prompt building

### Seams check (for implementer / reviewer)

The intended testing focus is **one primary seam** (RCA graph invoke → enriched record shape). Secondary seams cover predictor API and detector output only where the graph test cannot substitute. Confirm this matches expectations before implementation.

## Out of Scope

- Full MLflow registry UI demo and automated promote-to-production CI (support URI loading; full pipeline is a follow-on demo).
- KServe Predict V2 protocol (`/v2/models/...:infer`) — custom container + `/v1/classify` only.
- Retraining notebook changes beyond exporting a compatible artifact.
- Frequency per timestep on `kpi_window` (band on parent is sufficient).
- GPU ServingRuntime configuration (CPU inference acceptable for demo scale).
- Changes to `ran-anomaly-detector` rule logic or new anomaly types.
- Edge infrastructure workflow (`agent-service` graph).
- AutoRAG or Granite model serving changes.
- Chatbot UI beyond context string formatting.

## Further Notes

- Parent Jira: APPENG-5050 (Telco Network Automation quickstart). Sub-task: APPENG-6062.
- Based on Alan’s Mantis/TelecomTS work (Google doc in Jira).
- When ML is disabled (`MANTIS_ENABLED=false`) or inference URL unset, entire classify path is no-op — same behavior as today.
- `ml_root_cause_class` is the **ML category label**; `root_cause` remains the **LLM explanation** — they may disagree by design when steering is gated.
- Cold start: partial windows below `MANTIS_MIN_WINDOW` → LLM-only until detector buffer fills.
- Demo narrative beat: MLflow model → InferenceService on OpenShift AI → Classify node → LangGraph RAG + Analyze → enriched topic → chatbot shows both ML and LLM outputs.
